crates/red_knot_python_semantic/resources/mdtest/generics/pep695.md

```diff
@@ -48,4 +48,120 @@ class C[T]:
         reveal_type(x)  # revealed: T
 ```
 
+## Subtyping and assignability
+
+An unbounded, unconstrained typevar is assignable to itself, but is not a subtype of itself (or any
+other type), since it might be specialized to `Any`, which does not participate in the subtyping
+relationship.
+
+It is neither assignable to or a subtype of any other type (including other typevars), since we can
+make no assumption about what type it will be specialized to.
+
+```py
+from knot_extensions import is_assignable_to, is_subtype_of, static_assert
+
+def unbounded_unconstrained[T, U](t: T, u: U) -> None:
+    static_assert(is_assignable_to(T, T))
+    static_assert(is_assignable_to(U, U))
+    static_assert(not is_assignable_to(T, U))
+    static_assert(not is_assignable_to(U, T))
+
+    static_assert(not is_subtype_of(T, T))
+    static_assert(not is_subtype_of(U, U))
+    static_assert(not is_subtype_of(T, U))
+    static_assert(not is_subtype_of(U, T))
+```
+
+A bounded typevar is assignable to its bound, but the bound is not assignable to the typevar (since
+the typevar might be specialized to a smaller type). (TODO: Unless the bound is final, in which case
+the final class is also assignable to the typevar.)
+
+```py
+from typing_extensions import final
+
+def bounded[T: int](t: T) -> None:
+    static_assert(is_assignable_to(T, int))
+    static_assert(not is_assignable_to(int, T))
+
+    static_assert(not is_subtype_of(T, int))
+    static_assert(not is_subtype_of(int, T))
+
+@final
+class FinalClass: ...
+
+def bounded_final[T: FinalClass](t: T) -> None:
+    static_assert(is_assignable_to(T, FinalClass))
+    # TODO: is_assignable_to
+    static_assert(not is_assignable_to(FinalClass, T))
```

Comment:
See above: I'm not sure if this should really be assignable. `FinalClass` can have no subclasses, but it does have subtypes. So `T` could be `Never`, and  `FinalClass` is not assignable to `T`.


Unrelated hint: I usually encode the assertion that I *want* to hold true and silence the static-assertion error using `# error`. This way, the TODO can be resolved by removing the error assertion instead of having to invert the Boolean condition. It's really just a stylistic preference, but I think it makes the tests a bit easier to read if you can just ignore the comments and see the actual assertions in code:
```suggestion
    # TODO: This should hold true
    # error: [static-assert-error]
    static_assert(is_assignable_to(FinalClass, T))
```

Reply:
Maybe it could be argued for this particular function that if `T` is `Never`, the function could never be called (as there is no value which could be passed in as argument). And in this sense, `FinalClass` could be considered assignable to `T`, since the body would not be executed if `T` would be `Never`.

But then I could change the example slightly to
```py
def bounded_final[T: FinalClass](t: list[T]) -> T:
    # same body
```
and this function *can* be called with an empty list of type `list[Never]`. And therefore I would conclude that `FinalClass` should not be assignable to `T`.

Reply:
> Unrelated hint: I usually encode the assertion that I _want_ to hold true and silence the static-assertion error using `# error`.

I like that, thank you!

Reply:
> But then I could change the example slightly to

I updated all of the functions to use `list[?]` for consistency

> Unrelated hint: I usually encode the assertion that I _want_ to hold true and silence the static-assertion error using `# error`.

With the changes to how we're handling final bounds/constraints, it turns out there aren't any more TODOs.  But I will keep this in mind for the future!